staging/src/k8s.io/apimachinery/pkg/util/validation/field/testing/match.go

```diff
@@ -138,66 +98,104 @@ func (m *Matcher) Render(e *field.Error) string {
 	m.matchField = true
 	return m
 }
 
-// ByValue configures the matcher to match errors by the errant value.
-func (m *Matcher) ByValue() *Matcher {
+// ByValue returns a derived ErrorMatcher which also matches by the errant
+// value.
+func (m ErrorMatcher) ByValue() ErrorMatcher {
 	m.matchValue = true
 	return m
 }
 
-// ByOrigin configures the matcher to match errors by the origin.
-func (m *Matcher) ByOrigin() *Matcher {
+// ByOrigin returns a derived ErrorMatcher which also matches by the origin.
+func (m ErrorMatcher) ByOrigin() ErrorMatcher {
 	m.matchOrigin = true
 	return m
 }
 
-// RequireOriginWhenInvalid configures the matcher to require the Origin field
-// to be set when the Type is Invalid and the matcher is matching by Origin.
-func (m *Matcher) RequireOriginWhenInvalid() *Matcher {
+// RequireOriginWhenInvalid returns a derived ErrorMatcher which also requires
+// the Origin field to be set when the Type is Invalid and the matcher is
+// matching by Origin.
+func (m ErrorMatcher) RequireOriginWhenInvalid() ErrorMatcher {
 	m.requireOriginWhenInvalid = true
 	return m
 }
 
-// ByDetailExact configures the matcher to match errors by the exact detail
-// string.
-func (m *Matcher) ByDetailExact() *Matcher {
+// ByDetailExact returns a derived ErrorMatcher which also matches errors by
+// the exact detail string.
+func (m ErrorMatcher) ByDetailExact() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return got == want
 	}
 	return m
 }
 
-// ByDetailSubstring configures the matcher to match errors by a substring of
-// the detail string.
-func (m *Matcher) ByDetailSubstring() *Matcher {
+// ByDetailSubstring returns a derived ErrorMatcher which also matches errors
+// by a substring of the detail string.
+func (m ErrorMatcher) ByDetailSubstring() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return strings.Contains(got, want)
 	}
 	return m
 }
 
-// ByDetailRegexp configures the matcher to match errors by a regular
-// expression of the detail string, where the "want" string is assumed to be a
-// valid regular expression.
-func (m *Matcher) ByDetailRegexp() *Matcher {
+// ByDetailRegexp returns a derived ErrorMatcher which also matches errors by a
+// regular expression of the detail string, where the "want" string is assumed
+// to be a valid regular expression.
+func (m ErrorMatcher) ByDetailRegexp() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return regexp.MustCompile(want).MatchString(got)
 	}
 	return m
 }
+
+// Test compares two ErrorLists by the criteria configured in this matcher, and
+// fails the test if they don't match. If a given "want" error matches multiple
+// "got" errors, they will all be consumed. This might be OK (e.g. if there are
+// multiple errors on the same field from the same origin) or it might be an
+// insufficiently specific matcher, so these will be logged.
+func (m ErrorMatcher) Test(t *testing.T, want, got field.ErrorList) {
```

Comment:
Nit: `tb testing.TB` makes this slightly more versatile.

There's no immediate need for using this in a benchmark, but who knows...

Reply:
I saw you did that in some of the DRA tests and didn't really know why.  Is the theory that this might use the B part?  I'm not sure I follow

Reply:
This is a generic test helper. Benchmarks are also tests and might want to use the same helper to check some assertions. They can't if the helper expects a `testing.T`.

This might not be necessary here, but if it ever is, then it will be annoying to go back and have to ask for approval to change `testing.T` into `testing.TB` and rename the variable, or use `t testing.TB` to minimize the churn.

Reply:
Ahh, TB is an interface while T is not.  Sure.